simulation/agent.py

```diff
@@ -179,11 +187,50 @@ def _update_sprint(self):
         #Check, if the agent has rested for enough -> ensure, that rests when if can't sprint
         if self.is_resting:
             self.move_speed = 0
+            
+    def update_tower(self):
+        #The agent has to take N seconds to interact with the tower
+        if (self._world.time_ticks - self._tower_start_time) < (self._tower_interacction_time/self._world.TIME_PER_TICK):
+            return
+        
+        #Sanity check. Return if the agent is not interacting with a tower
+        if not self._interacting_with_tower:
+            return 
+        
+        #Leaving the tower
+        if self._in_tower:
+            #Move the agent out of the tower range in the direction that he is heading
+            self.move_speed = self.base_speed
+            self.move(self._width*1.2)
+            #TODO: Set vision range to normal (left tower)
+
+        self._in_tower = not self._in_tower
+        self._is_deaf = False
+        self._interacting_with_tower = False
+        #TODO: Set vision range to [2,30] (in the tower)
+        
+                                
+    def interact_with_tower(self, tower):
+        if not self.in_tower_range(tower) or self._interacting_with_tower:
+            return
+        #Put agent on tower and deafen him
+        self._interacting_with_tower = True
+        self.location.x = tower[0]+self._width/2
+        self.location.y = tower[1]+self._width/2
+        self.move_speed = 0
+        self._tower_start_time = self._world.time_ticks
+        self._is_deaf = True
+
+    def in_tower_range(self, tower):
+        distance = math.sqrt((tower[0]-self.location.x)**2+(tower[1]-self.location.y)**2)
```

Comment:
this can just be done using the vector operations, something like `(tower_location - self.location).length` IIRC

Reply:
You are the python pro here :D
I wasn't aware of that but it surely makes it more readable

Reply:
It's not a python thing, it's because I'm using the `vectormath` library :)